engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java

```diff
@@ -44,12 +46,14 @@ protected void handleInnerActivity(final BpmnStepContext<ExecutableMultiInstance
 
     if (loopCharacteristics.isSequential()) {
 
-      final var array = readInputCollectionVariable(context).getSingleResult().getArray();
+      final var array =
+          readInputCollectionVariable(context).orElseThrow(IllegalStateException::new);
```

Comment:
Throwing an exception here is dangerous :boom: It is very unlikely but the input collection variable might change. If it is not an array anymore then an exception is thrown and the workflow instance stops processing at all (i.e. added to the "blacklist").

It would be better to handle the case and create an incident instead. So, the workflow instance can be repaired and continue processing (which is not the case if it is added to the "blacklist").

Reply:
@saig0 Ah okay, that's a good reason to avoid throwing an exception here. But how should we deal with the situation then? In case the variable is no longer an array, an incident is already raised inside the `readInputCollectionVariable` method. So we don't have to do that here. However, should we immediately return from this `handleInnerActivity` when we see the optional to be empty or should the rest of the method still be executed? For example, do we still need to call `super.handleInnerActivity`?

Reply:
:+1: We should return `false` immediately. Otherwise, we may continue the workflow instance. 

I have more and more the feeling that this API needs an `Either` return type...

Reply:
@saig0 I'm all for it. Here are some implementations for Java: https://stackoverflow.com/questions/26162407/is-there-an-equivalent-of-scalas-either-in-java-8. The only issue I see with introducing one to our code is consistency. I think the ExpressionProcessor API should be migrated to using it in one PR, to make sure how we use it in the same way everywhere. That might take some effort.

Reply:
:+1: